### PR TITLE
Support for IPv6

### PR DESCRIPTION
**Pull request by aramase**
Add support for ipv6.

This is to add support for the dual stack feature - 
https://github.com/kubernetes/kubernetes/pull/78801
https://github.com/kubernetes/kubernetes/pull/79386

cc @khenidak

### CODE REVIEW

**Comment by khenidak** [comment]
/assign @thockin 

This is for masquerading, as we discussed last week

**Comment by aojea** [comment]
Fixes: https://github.com/kubernetes-incubator/ip-masq-agent/issues/43

**Comment by aojea** [comment]
@aramase I think that this will fail on systems that don't have ipv6 or ipv4 enabled, it's common to have both enabled but we should consider systems with only one protocol enabled.
 In addition, I think that the user experience will be better if it can configure the ipv4 and ipv6 CIDR independently..
I shared some thoughts in https://github.com/kubernetes-incubator/ip-masq-agent/issues/43 ,

**Comment by aramase** [comment]
@aojea Thank you for taking a look at this. The configuration takes in a CIDR list, so I thought it might be easier for users to just provide a list of all the CIDRs (ipv4/ipv6). I'm thinking maybe we can add an agent flag to enable ipv6? Thoughts?

**Comment by aojea** [comment]
> @aojea Thank you for taking a look at this. The configuration takes in a CIDR list, so I thought it might be easier for users to just provide a list of all the CIDRs (ipv4/ipv6). 

you are totally right ... also, this way the configuration is backward compatible

> I'm thinking maybe we can add an agent flag to enable ipv6? By default that will be set to false. Thoughts?

I think that´s a good idea, my thoughts are ...
Users should be able to upgrade without any changes, what´s the best option? using a flag or autodetecting it from the configuration, ... ?
Users should be able to use the agent for IPv4 only, IPv6 only or dual-stack mode

Regarding dual stack I have one doubt, what should happen if one of the protocols fails, let says that someone uses dual stack and the system doesn´t have the IPv6 tools and can´t install the ipv6 rules, should the agent fail or keep working with the IPv4 iptables rules?

**Comment by khenidak** [comment]
I think best is to add flag for `ipv6` if enabled we non masq the link local. For the user provided list we don't anything, just a list of CIDRs.

**Comment by aramase** [comment]
> I think best is to add flag for ipv6 if enabled we non masq the link local. For the user provided list we don't anything, just a list of CIDRs.

@khenidak If we add a flag, is the suggestion that we masquerade everything and if `--enable-ipv6` then we non-masquerade link local? No changes to cluster-cidr makes sense. If the user provides an ipv6 in the list, then we add that in iptables.

**Comment by bowei** [comment]
cc: @varunmar

File: cmd/ip-masq-agent/ip-masq-agent.go [hunk 1]
```diff
@@ -240,6 +246,8 @@ func validateCIDR(cidr string) error {
 func (m *MasqDaemon) syncMasqRules() error {
 	// make sure our custom chain for non-masquerade exists
 	m.iptables.EnsureChain(utiliptables.TableNAT, masqChain)
+	// make sure our custom chain for ipv6 non-masquerade exists
+	m.ip6tables.EnsureChain(utiliptables.TableNAT, masqChain)
```
Comment (bowei) [inline cmd/ip-masq-agent/ip-masq-agent.go]: As per the discussion -- should we call EnsureChain if there is no ipv6 IPs?
(This applies to the other calls to ip6tables as well

File: cmd/ip-masq-agent/ip-masq-agent_test.go [hunk 2]
```diff
@@ -294,19 +358,38 @@ func TestEnsurePostroutingJump(t *testing.T) {
...
+			want: string(utiliptables.Append) + " " + string(masqChain) +
+				` -m comment --comment "ip-masq-agent: local traffic is not subject to MASQUERADE"` +
+				" -d fc00::/7 -j RETURN\n",
+		},
 	}
-	if s != want {
-		t.Errorf("writeNonMasqRule(lines, "+cidr+"):\n   got: %q\n  want: %q", s, want)
+
+	for _, tt := range writeNonMasqRuleTests {
+		lines := bytes.NewBuffer(nil)
```
Comment (bowei) [inline cmd/ip-masq-agent/ip-masq-agent_test.go]: t.Run(tc.desc, ...

**Comment by varunmar** [comment]
I'd like a way to ensure that today's behavior can be replicated exactly after this change goes in.

Ideally, this would be the default if exactly the same flags and configs were used today and version n+1.
I'd be okay with it if we had to explicitly specify a flag to return to the old behavior.
I'd be much less happy if it also required a change in the default config (many customers run this addon with no additional config, it would be unfortunate if we forced them all to manage another configmap before they are ready)

I think @khenidak's suggestion is a good one - a flag named --enable-ipv6 that is false by default. If true, then write the link-local ip6 rule if the configmap is empty. If a list of CIDRs is present, then use them instead.

So now the only question is what to do if the --enable-ipv6 flag is false and a configmap with ipv6 addresses is passed in?
Another question to consider is what to do if IPv6 tools are not installed on the machine (or less commonly, no ipv4 tools)

I think the answer to both should be that an error condition should be raised and the daemon must not report healthy status. The easiest way may be to turn it into an infinite crash loop till the config is fixed, but maybe there are better solutions.

Here's what I think the behavior should be:

Configuration | Behavior
-|-
--enable-ipv6=false, no CIDR entries | no change from today's behavior (no-masq default set of ipv4 cidrs)
--enable-ipv6=false, CIDR entries contain ipv4 only | only no-masq the CIDR entries explicitly
--enable-ipv6=true, no CIDR entries | no-masq default set of ipv4 + no-masq link-local ipv6
--enable-ipv6=true, CIDR entries contain ipv4 and/or ipv6 | no-masq CIDR entries only
--enable-ipv6=true, CIDR entries contain ipv4 only | no-masq CIDR entries only (do not no-masq ipv6 link-local)
--enable-ipv6=false, CIDR entries contain ipv6 entries | Error, crash
--enable-ipv6=true, no ipv6 or ip6tools on machine | Error, crash

File: cmd/ip-masq-agent/ip-masq-agent.go [hunk 3]
```diff
@@ -311,3 +340,18 @@ func writeRule(lines *bytes.Buffer, position utiliptables.RulePosition, chain ut
...
+// If cidr belongs to ipv6 family, return true else it returns false
+// which means the cidr belongs to ipv4 family
+func isIPv6CIDR(cidr string) bool {
+	ip, _, _ := net.ParseCIDR(cidr)
+	return isIPv6(ip.String())
+}
+
+// isIPv6 checks if the provided ip belongs to ipv6 family.
+// If ip belongs to ipv6 family, return true else it returns false
+// which means the ip belongs to ipv6 family
```
Comment (varunmar) [inline cmd/ip-masq-agent/ip-masq-agent.go]: typo - "ipv4 family" :)

**Comment by aramase** [comment]
> I'd be okay with it if we had to explicitly specify a flag to return to the old behavior.

By introducing the `--enable-ipv6` agent flag we will retain the old behavior. Default will be false which means it'll only setup the ipv4 chain and rules.

> I'd be much less happy if it also required a change in the default config (many customers run this addon with no additional config, it would be unfortunate if we forced them all to manage another configmap before they are ready)

Totally understand. We shouldn't change the default config.

> I think @khenidak's suggestion is a good one - a flag named --enable-ipv6 that is false by default. If true, then write the link-local ip6 rule if the configmap is empty. If a list of CIDRs is present, then use them instead.

I think instead of checking only if the configmap with cidr exists, maybe we should check if ipv6 cidrs exists in the list? If not, then we should still write the link-local ip6 rule if ipv6 is enabled. 

Totally agree with the behavior you suggested with just few changes -


Configuration | Behavior
-- | --
--enable-ipv6=false, no CIDR entries | no change from today's behavior (no-masq default set of ipv4 cidrs, do not no-masq ipv6 link-local)
--enable-ipv6=false, CIDR entries contain ipv4 only | no-masq the CIDR entries only
--enable-ipv6=true, no CIDR entries | no-masq default set of ipv4 + no-masq link-local ipv6
--enable-ipv6=true, CIDR entries contain ipv4 only | no-masq CIDR entries only + no-masq link-local ipv6
--enable-ipv6=true, CIDR entries contain ipv4 and ipv6 | no-masq CIDR entries only (do not no-masq ipv6 link-local)
--enable-ipv6=false, CIDR entries contain ipv6 entries | Error, crash
--enable-ipv6=true, no ipv6 or ip6tools on machine | Error, crash

Do you think this behavior is acceptable? @varunmar @khenidak

**Comment by aramase** [comment]
@bowei @varunmar I've added the flag and updated the PR. PTAL.

File: README.md [hunk 4]
```diff
@@ -28,6 +28,7 @@ By default, the agent is configured to reload its configuration from the `/etc/c
 The agent configuration file should be written in yaml or json syntax, and may contain three optional keys:
 - `nonMasqueradeCIDRs []string`: A list strings in CIDR notation that specify the non-masquerade ranges.
 - `masqLinkLocal bool`: Whether to masquerade traffic to `169.254.0.0/16`. False by default.
+- `masqLinkLocalIPv6 bool`: Whether to masquerade traffic to `fe80::/10`. False by default.
```
Comment (varunmar) [inline README.md]: May be worth mentioning that this has no effect unless enable-ipv6 is true.

**Comment by aramase** [comment]
@bowei Can you PTAL at this PR when you get a chance?

**Comment by thockin** [comment]
I am buried, so I am assigning to Varun.  @varunmar if you don't want it, let me know.

**Comment by aramase** [comment]
@thockin Thank you! @varunmar has reviewed the PR and added `/lgtm` but looks like a `/lgtm` is required from a collaborator? https://github.com/kubernetes-incubator/ip-masq-agent/pull/45#issuecomment-511584713

File: cmd/ip-masq-agent/ip-masq-agent_test.go [hunk 5]
```diff
@@ -283,6 +314,68 @@ COMMIT
...
+// tests MasqDaemon.syncMasqRulesIPv6
+func TestSyncMasqRulesIPv6(t *testing.T) {
+	var syncMasqRulesIPv6Tests = []struct {
+		desc string      // human readable description of the test
+		cfg  *MasqConfig // Masq configuration to use
+		err  error       // expected error, if any. If nil, no error expected
+		want string      // String expected to be sent to iptables-restore
+	}{
+		{
+			desc: "empty config",
```
Comment (bowei) [inline cmd/ip-masq-agent/ip-masq-agent_test.go]: this feels like we should use text template here for readability.

**Pull request merged**